### PR TITLE
Backup types

### PR DESCRIPTION
**Pull request by moelsayed**

### CODE REVIEW

File: apis/management.cattle.io/v3/backup_types.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import (
+	"github.com/rancher/norman/condition"
+	"github.com/rancher/norman/types"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	BackupConditionCreated    condition.Cond = "Created"
+	BackupConditionCompeleted condition.Cond = "Compeleted"
```
Comment (alena1108) [inline apis/management.cattle.io/v3/backup_types.go]: typo in completed

File: apis/management.cattle.io/v3/backup_types.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+)
+
+type BackupBackend struct {
+	// backend name
+	Name string `yaml:"name" json:"name,omitempty" norman:"default=s3"`
```
Comment (alena1108) [inline apis/management.cattle.io/v3/backup_types.go]: Lets think if we really need a name...

File: apis/management.cattle.io/v3/backup_types.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	// s3 backend
+	S3BackupBackend *S3BackupBackend `yaml:",omitempty" json:"s3BackupBackend"`
+}
+
+type S3BackupBackend struct {
+	// Access key ID
+	AccessKeyID string `yaml:"access_key_id" json:"accessKeyId,omitempty"`
+	// Secret access key
+	SecretAccesssKey string `yaml:"secret_access_key" json:"secretAccessKey,omitempty"`
```
Comment (alena1108) [inline apis/management.cattle.io/v3/backup_types.go]: should add norman annotation `password`

File: apis/management.cattle.io/v3/backup_types.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	// name of the bucket to use for backup
+	BucketName string `yaml:"bucket_name" json:"bucketName,omitempty"`
+	// AWS Region, AWS spcific
+	Region string `yaml:"region" json:"region,omitempty"`
+	// Endpoint is used if this is not an AWS API
+	Endpoint string `yaml:"endpoint" json:"endpoint"`
+}
```
Comment (alena1108) [inline apis/management.cattle.io/v3/backup_types.go]: if you want custom name for the backups, you can add prefix as a field here

File: apis/management.cattle.io/v3/backup_types.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+type BackupTarget struct {
+	// s3 target
+	S3BackupTarget *S3BackupTarget `yaml:",omitempty" json:"s3BackupTarget"`
+}
+
+type S3BackupTarget struct {
+	// Access key ID
+	AccessKeyID string `yaml:"access_key_id" json:"accessKeyId,omitempty"`
+	// Secret access key
+	SecretAccesssKey string `yaml:"secret_access_key" json:"secretAccessKey,omitempty" norman:"required,type=password" `
```
Comment (vincent99) [inline apis/management.cattle.io/v3/backup_types.go]: JSON/YAML: We call these `accessKey` and `secretKey` (`access_key` / `secret_key`) in other places of the API (EC2 and EKS)

File: apis/management.cattle.io/v3/backup_types.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	Endpoint string `yaml:"endpoint" json:"endpoint"`
+}
+type EtcdBackup struct {
+	types.Namespaced
+
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// actual file name on the target
+	FileName string `yaml:"file_name" json:"fileName,omitempty"`
```
Comment (vincent99) [inline apis/management.cattle.io/v3/backup_types.go]: JSON/YAML: `filename`

File: apis/management.cattle.io/v3/rke_types.go [hunk 7]
```diff
@@ -183,6 +183,8 @@ type ETCDService struct {
 	Retention string `yaml:"retention" json:"retention,omitempty" norman:"default=72h"`
 	// Etcd snapshot Creation period
 	Creation string `yaml:"creation" json:"creation,omitempty" norman:"default=12h"`
```
Comment (vincent99) [inline apis/management.cattle.io/v3/rke_types.go]: It thought we were going to fix this and Retention to be an integer number of minutes?  Is that a future release?

Don't like string times with complicated parsing and validation.

Comment (moelsayed) [reply apis/management.cattle.io/v3/rke_types.go]: We will move this to a future iteration. For now, UI can set this to a number of hours.

File: apis/management.cattle.io/v3/backup_types.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+type BackupTarget struct {
+	// s3 target
+	S3BackupTarget *S3BackupTarget `yaml:",omitempty" json:"s3BackupTarget"`
+}
+
+type S3BackupTarget struct {
+	// Access key ID
+	AccessKey string `yaml:"access_key" json:"accessKey,omitempty"`
+	// Secret access key
+	SecretKey string `yaml:"secret_key" json:"secretKey,omitempty" norman:"required,type=password" `
```
Comment (cjellick) [inline apis/management.cattle.io/v3/backup_types.go]: @kinarashah will this be stored as a k8s secret out of the gate?

File: apis/management.cattle.io/v3/schema/schema.go [hunk 9]
```diff
@@ -178,6 +180,10 @@ func clusterTypes(schemas *types.Schemas) *types.Schemas {
 				Input: "monitoringInput",
 			}
 			schema.ResourceActions["disableMonitoring"] = types.Action{}
+			schema.ResourceActions["saveSnapshot"] = types.Action{}
```
Comment (alena1108) [inline apis/management.cattle.io/v3/schema/schema.go]: Lets change it to `backupEtcd`

File: apis/management.cattle.io/v3/rke_types.go [hunk 10]
```diff
@@ -185,6 +187,8 @@ type ETCDService struct {
...
+	// Backup backend for etcd snapshots, used by rke only
+	BackupTarget *BackupTarget `yaml:"backup_target" json:"backupTarget,omitempty"`
```
Comment (alena1108) [inline apis/management.cattle.io/v3/rke_types.go]: Lets rename to the `BackupConfig`

File: apis/management.cattle.io/v3/cluster_types.go [hunk 11]
```diff
@@ -231,6 +231,11 @@ type IngressCapabilities struct {
 type MonitoringInput struct {
 	Answers map[string]string `json:"answers,omitempty"`
 }
+
+type RestoreSnapshotInput struct {
+	SnapshotName string        `json:"snapshotName,omitempty" norman:"type=reference[etcdBackup]"`
```
[the remark this answers is not in the document]
Comment (alena1108) [reply apis/management.cattle.io/v3/cluster_types.go]: lets rename it to etcdBackupName

File: apis/management.cattle.io/v3/rke_types.go (hunk 7 above)
Comment (alena1108) [reply apis/management.cattle.io/v3/rke_types.go]: update: we are introducing retention and intervalHours parameters to the BackupConfig as per conversation with @vincent99 : https://github.com/rancher/types/pull/633/commits/487bca05840ccf16008babd4ade145541e6ce803#diff-623d9a24d8befd392620d1315c53f7d2R18

File: apis/management.cattle.io/v3/backup_types.go (hunk 8 above)
Comment (kinarashah) [reply apis/management.cattle.io/v3/backup_types.go]: Yeah, have a pr coming in with some minor changes, with that it should work out of the gate

File: apis/management.cattle.io/v3/backup_types.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+	// name of the bucket to use for backup
+	BucketName string `yaml:"bucket_name" json:"bucketName,omitempty"`
+	// AWS Region, AWS spcific
+	Region string `yaml:"region" json:"region,omitempty"`
+	// Endpoint is used if this is not an AWS API
+	Endpoint string `yaml:"endpoint" json:"endpoint"`
+}
+type EtcdBackup struct {
+	types.Namespaced
```
Comment (alena1108) [inline apis/management.cattle.io/v3/backup_types.go]: we need to add a clusterid as a field

Comment (alena1108) [reply apis/management.cattle.io/v3/backup_types.go]: This field is so the UI can filter out the etcd backups on per cluster. And on the rancher api side, we should error out if we try to recover to the backups from other clusters.

File: apis/management.cattle.io/v3/rke_types.go (hunk 7 above)
Comment (alena1108) [reply apis/management.cattle.io/v3/rke_types.go]: @guangbochen We wanted to introduce a new format for these 2 parameters (hours for creation, number of snapshots for retention). But in order to not break api compatibility with the previous versions, we couldn't change the existing params, so we've introduced new ones in backupConfig, and those params would be applied to S3 backups (for now) only. 

> If both ETCDService and BackupConfig has Retention set with 12h and 6(num of backups), which one it should follow when removing the local backups

Old parameters should apply to the local backups only; new params from backupConfig should apply to the backups on S3.

Let me know if everything is clear @guangbochen 

 @moelsayed please confirm

Comment (guangbochen) [reply apis/management.cattle.io/v3/rke_types.go]: sure thanks @alena1108, I will add the s3 `backupConfig` to the rke-tools.

Comment (alena1108) [reply apis/management.cattle.io/v3/rke_types.go]: @guangbochen we don't need to add it to the rke tools; we just have to change rke side code to parse the correct retention and creation parameters.

**Comment by alena1108** [comment]
LGTM

**Pull request merged**